### ISSUE TITLE
:latest Docker version in Docker Hub is still 1.8.3

### ISSUE THREAD

**Issue by eburgueno** [oushujun/EDTA#122]
This is related to https://github.com/Kapeel/edta_docker/issues/1. The image currently tagged as "latest" in Docker Hub does not include version 1.9.0 as one would expect. It has 1.8.3 instead.

The likely reason is that since the `Dockerfile` does not specify the version of EDTA to install from Bioconda, the build server reused the cache for the previous container build.

Would you be interested in providing both Dockerfile and Singularity definition file(s) in this repo, or would you prefer to continue keeping them separate? I'll be happy to contribute a PR.

**Comment by oushujun** [comment]
@kapeel Can you help with this issue? Thanks! - Shujun

**Comment by oushujun** [comment]
@eburgueno Thanks for pointing out the version issue. You are also welcome to contribute a docker/singularity file(s) to this repo. If possible please also indicate how to install the program using this file(s). Thanks! -Shujun

**Comment by eburgueno** [comment]
Looking at this more closely, you probably don't need either Docker nor Singularity recipes. BioConda [automatically creates](https://biocontainers-edu.readthedocs.io/en/latest/conda_integration.html) a [Docker container in Quay.io](https://quay.io/repository/biocontainers/edta?tab=tags) for you based on your Conda recipe, so ideally you would just focus on keeping that up to date.

**Comment by oushujun** [comment]
Thanks for the suggestions! I am aware of this auto-build docker/singularity version before, but it had some issues dealing with the RepeatMasker dependency. I am not sure if it has improved, so I tried to build the suggested singularity version again. The build was successful but when I try to use it, I encounter the following warnings and errors:
>WARNING: Skipping mount /opt/rit/spack-app/linux-rhel7-x86_64/gcc-4.8.5/singularity-3.5.2-ewuuq5o5hv4r6lilhim47o5xzcu6cx4t/var/singularity/mnt/session/etc/resolv.conf [files]: /etc/resolv.conf doesn't exist in container
perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
        LANGUAGE = (unset),
        LC_ALL = (unset),
        LANG = "en_US.UTF-8"
    are supported and installed on your system.
perl: warning: Falling back to the standard locale ("C").

>Sun Oct 18 23:14:30 CDT 2020    Dependency checking:
Error: The RMblast engine is not installed in RepeatMasker!
Command exited with non-zero status 30

The RepeatMasker bug is still there. Please refer to the [docker file](https://github.com/Kapeel/edta_docker/blob/master/Dockerfile) on how we bypass this bug.

The docker container has the following errors:
`docker pull quay.io/biocontainers/edta:1.9.0--0`
`docker run -v $PWD:/in -w /in quay.io/biocontainers/edta:1.9.0--0 EDTA.pl --genome genome.fa --cds genome.cds.fa --curatedlib rice6.9.5.liban --exclude genome.exclude.bed --sensitive 1 --anno 1 --evaluate 1 --threads 36 > EDTA.test0`

>sh: timeout: not found

This error is new. It seems like the `timeout` linux command used in the program is not included in the auto-build container. 

You can find more discussions on docker containers here: https://github.com/oushujun/EDTA/issues/105

**Comment by eburgueno** [comment]
Sorry it's taken me a while to get back to you. I understand why those are important to resolve before we can consider using biocontainers.

The `timeout` command is part of `coreutils`, which is not included in the final image because of the unique way biocontainers are created. But `coreutils` has a [conda-forge package](https://anaconda.org/conda-forge/coreutils/), so if we add that to the list of dependencies that might just fix that problem. As a bonus, in our tests I also noticed you make use of `sort` with the `-V` flag. The biocontainer image is based on busybox which provides a minimal implementation of `sort` that does not work as intended. Installing `coreutils` takes care of this issue as well.

The RMblast engine error is something I encountered as well. My solution was slightly simpler:
```
cd /opt/conda/share/RepeatMasker
perl ./configure -trfbin=`which trf` -hmmerbin=`which nhmmscan` -rmblastbin=`which rmblastn`
```

I am not sure if you are directly involved in managing the Bioconda package, but it sounds like this could be added to the [`build.sh` recipe](https://github.com/bioconda/bioconda-recipes/blob/master/recipes/edta/build.sh) so both the final conda environment as well as the resulting container images would be functional out of the box.

Lastly, I noticed that you have bumped the version to 1.9.3 but there's no new GitHub release. You will probably need to create one to trigger the Bioconda build bots.

I understand none of this is exciting from a science point of view. Release management is definitely painful. Hopefully the above suggestions make it slightly less frustrating.

**Comment by oushujun** [comment]
Thank you for all these great suggestions! I think your solution to installing RepeatMasker is elegant and I like it! Software maintenance is not trivial but it's important for the entire community. Seeing joys from users excite me as much as new scientific findings. I think the TE community is getting better and better with people contributing their feedback, ideas, and solutions. 

@Juke34 Can you incorporate `coreutils` and the `RepeatMasker` solution to the bioconda recipe? I will push some updates and release the v1.9.4 and hopefully that solves both conda and docker issues.

**Comment by Juke34** [comment]
I can include coreutils in the EDTA recipe but I don't want to add a fix for a dependency (i.e. RepeatMasker). I would prefer to fix the RepeatMasker recipe directly if possible.

**Comment by Juke34** [comment]
Coreutils added see here: https://github.com/bioconda/bioconda-recipes/pull/25089

**Comment by Juke34** [comment]
RepeatMasker has been updated to version 4.1.1 and should work fine now https://github.com/bioconda/bioconda-recipes/pull/25163

**Comment by oushujun** [comment]
@Juke34 I just got a chance to tested the biocontainers version of EDTA. The new RepeatMasker conda recipe works pretty well in the EDTA docker/singularity container. However I encounter two other issues:

Issue 1:
>perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
        LANGUAGE = "en_US:",
        LC_ALL = (unset),
        LANG = "en_US.UTF-8"
    are supported and installed on your system.
perl: warning: Falling back to the standard locale ("C").

Same issue reported [here](https://github.com/BioContainers/containers/issues/206). This can be simply fixed by `export LC_ALL=C` before running the container. Is that a way to set this within the conda recipe?

Issue 2:

>b'\nError: Failed to open binary auxfiles for /usr/local/lib/python3.7/site-packages/TEsorter/database/REXdb_protein_database_
viridiplantae_v3.0_plus_metazoa_v3.hmm: use hmmpress first\n\n'

Same issue reported #121. I see TEsorter has been updated to v1.3.0, and it supposes to have this fixed? I double-check the conda version EDTA v1.9.4, there is no such error.

Thanks!
Shujun

**Comment by oushujun** [comment]
@Juke34 Thank you for adding `hmmpress` to the `TEsorter` conda recipe. I believe _Issue 2_ mentioned above will be solved upon the EDTA new release. 

While testing the docker version from quay.io, I notice another issue:

`/usr/local/share/EDTA/bin/TIR-Learner2.5/TIR-Learner2.5.sh: line 58: getopts: not found`

This is found in both macOS and Linux platforms. I check conda and found the Mac version of getopts: https://anaconda.org/bioconda/gnu-getopt. Not sure if including this package will solve the issue. Any ideas?

Thanks,
Shujun

**Comment by philippbayer** [comment]
/usr/local/share/EDTA/bin/TIR-Learner2.5/TIR-Learner2.5.sh: line 58: getopts: not found

I just bumped into this in a slightly different setting, I was using the github version but with a regular Ubuntu 18.04 singularity image. getopts is part of regular bash, but sometimes the EDTA_raw.pl calls on sh, not bash, and I guess there are some Ubuntu-based images out there where sh doesn't have getopts?

I changed line 387 into `bash $TIR_Learner -g $genome -s $species -t $threads -l $maxint` (so turned the call from sh to bash) and that seems to have fixed it, but I'm doublechecking

**Comment by oushujun** [comment]
@philippbayer It works! Both in macOS and centOS on my end. I will add this fix soon. Thanks.

**Comment by philippbayer** [comment]
Good to hear :)

**Comment by oushujun** [comment]
Updated to v1.9.5 but there is still a bug. See #137
